foundation/org.eclipse.persistence.core/src/main/java/org/eclipse/persistence/internal/queries/DatabaseQueryMechanism.java

```diff
@@ -873,6 +884,64 @@ protected void updateObjectAndRowWithSequenceNumber() throws DatabaseException {
         writeQuery.getDescriptor().getObjectBuilder().assignSequenceNumber(writeQuery);
     }
 
+    /**
+     * Update the object's primary key by getting the generated keys from the call
+     * If there are no generated keys or the value is NULL, then default back to the {@link #updateObjectAndRowWithSequenceNumber()}
+     */
+    protected void updateObjectAndRowWithSequenceNumber(DatabaseCall call) throws DatabaseException {
```

Comment:
Why do you need to have an instance of the DatabaseCall and cannot use the one obtained via `writeQuery.getCall()`? Would that allow one to get rid of this new method and the change in the return type?

Reply:
@lukasj Sorry, I have been a bit busy and didn't get to this

> Why it is required to have an instance of the DatabaseCall here instead of using writeQuery.getCall()?

The issue is that `writeQuery.getCall()` returns the QueryMechanism's Call object, which is the original and not the clone. Query's don't actually store Call objects.

It's all about execution Calls vs querymechanism Calls. The former is just a clone of the latter. You can see where this occurs in the stack here: `org.eclipse.persistence.internal.queries.DatasourceCallQueryMechanism.executeCall(DatasourceCall databaseCall)`

During `org.eclipse.persistence.internal.databaseaccess.DatabaseAccessor.basicExecuteCall(...)`, all I have available is a Call, no Query. But it's here, at this depth in the call stack, that we have the `java.sql.Statement` so we can get the GeneratedKeys ResultSet. This Call object is the execution clone. So, I am placing the GeneratedKeys ResultSet on the cloned Call and returning it up the stack. I think this makes sense functionally as Call's are basically just Action objects and the clone is disposable once the action is completed. 

The cloned Call then returns back up the stack, eventually arriving to `org.eclipse.persistence.internal.queries.DatasourceCallQueryMechanism.insertObject()` and passing it to `org.eclipse.persistence.internal.queries.DatabaseQueryMechanism.updateObjectAndRowWithSequenceNumber(DatabaseCall call). In order to use `writeQuery.getCall()`, I would need to set the ResultSet on the original Call object. Which I think is the wrong place to do that.